src/formik.tsx

```diff
@@ -217,75 +255,125 @@ Formik cannot determine which value to update. See docs for more information: ht
+        }
+      };
+
+      setFieldValue = (field: string, value: any) => {
+        // Set form field by name
+        this.setState(prevState => ({
+          ...prevState,
+          values: {
+            ...prevState.values as object,
+            [field]: value,
+          },
+        }));
+
+        if (validateOnChange) {
+          this.validateFormik({ ...this.state.values as any, [field]: value });
+        }
       };
 
       handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
         e.preventDefault();
-        // setTouched();
-        // setSubmitting(true);
         this.setState({
           touched: touchAllFields(this.state.values),
           isSubmitting: true,
         });
-        const { values } = this.state;
-        // Validate against schema
-        validateFormData<Values>(values, validationSchema).then(
+
+        validateFormData<Values>(this.state.values, validationSchema).then(
           () => {
             this.setState({ errors: {} });
             handleSubmit(mapValuesToPayload(this.state.values), {
               setTouched: this.setTouched,
               setErrors: this.setErrors,
               setError: this.setError,
-              setSubmitting: this.setSubmitting,
               setValues: this.setValues,
+              setFieldError: this.setFieldError,
+              setFieldValue: this.setFieldValue,
+              setFieldTouched: this.setFieldTouched,
+              setSubmitting: this.setSubmitting,
               resetForm: this.resetForm,
               props: this.props,
             });
           },
-          (err: any) => this.setState({ isSubmitting: false, errors: yupToFormErrors(err) })
+          (err: any) =>
+            this.setState({ isSubmitting: false, errors: yupToFormErrors(err) })
         );
       };
 
       handleBlur = (e: any) => {
         e.persist();
         const { name, id } = e.target;
         const field = name ? name : id;
-        this.setState(state => ({ touched: { ...state.touched, [field]: true } }));
+        this.setState(prevState => ({
+          touched: { ...prevState.touched, [field]: true },
+        }));
+
+        if (!validateOnChange) {
+          this.validateFormik(this.state.values);
+        }
       };
 
-      handleChangeValue = (field: string, value: any) => {
-        const { touched, values } = this.state;
-        // Set touched and form fields by name
-        this.setState(state => ({
-          ...state,
-          values: {
-            ...values as object,
-            [field]: value,
-          },
+      setFieldTouched = (field: string, touched: boolean = true) => {
+        // Set touched field by name
+        this.setState(prevState => ({
+          ...prevState,
           touched: {
-            ...touched as object,
-            [field]: true,
+            ...prevState.touched,
+            [field]: touched,
+          },
+        }));
+
+        // Async validate
+        if (!validateOnChange) {
```

Comment:
What if we add a `validateOnTouch` option so you can separately control whether you want a change, a touch, both, or neither, to validate?